[WFLY-20951] Take into account the new Jakarta Authorization capability.

darranl opened the pull request:
I am going to start as a draft but this is independently mergable without the rest.

Longer term we may want to introduce a Jakarta Authorization layer / extension - for now the Elytron extension will register Jakarta Authorization 3.0 for Jakarta EE 11 if the integration supports it and if so will record a capability:

https://github.com/darranl/wildfly-core/blob/WFCORE-7322/elytron/src/main/java/org/wildfly/extension/elytron/ElytronDefinition.java#L490-L497

This means we have a capability we can detect but we don't need to depend on a Service for EE 11.

https://issues.redhat.com/browse/WFLY-20951

darranl committed:
[WFLY-20951] Update the EAR deployment to take into accoun the new Jakarta Authorization capability.

darranl committed:
[WFLY-20951] Update the EJB deployment to take into accoun the new Jakarta Authorization capability.

darranl committed:
[WFLY-20951] Update the Undertow deployment to take into account the new Jakarta Authorization capability.

bstansberry commented:
@darranl I suggest a few words in the DUPs and in EeCapabilities noting the distinction between the 2 capabilities.

If that would be too wordy, then please spell it out clearly in the JIRA description.

bstansberry commented on ee/src/main/java/org/jboss/as/ee/security/JaccEarDeploymentProcessor.java:
```diff
@@ -27,10 +30,9 @@
 public class JaccEarDeploymentProcessor implements DeploymentUnitProcessor {
 
-    private final String jaccCapabilityName;
+    private volatile boolean deployed = false;
```
This needs a usage comment, because it's fragile. Maybe it needs a rename?

The instance of this class is not bound to a deployment, so any state it holds is globally applicable. But 'deployed' intuitively means something applicable to a particular deployment.

Maybe something like

// Tracks whether we've detected that one of the capabilities indicating a need for Jakarta Authorization is present
private volatile boolean detectedJakartaAuthentication = false;

The presence of absence of the capability won't change without a reload (otherwise this field isn't valid at all), so you could also change the 'deploy' impl to just skip looking for the capabilities if this is 'true'.

bstansberry replied:
+1. 

AbstractSecurityDeployer and subclasses should be renamed. They are AbstractJaccServiceFactory etc and have a 'createJaccService" method, not a 'deploy' method. They don't deploy anything in any sense other than adding an object to the heap.

That doesn't need to be in scope for this PR though.

darranl replied:
+1 I would actually like to see what we can pull out here, there is a little but too much code IMO in the WildFly repo.

If we could refactor and pull some out it would makeit easier to use with Undertow standalone as well which wouild also make developing and testing this a lot easier.

darranl committed:
[WFLY-20951] Add additional comments describing the relationship of the two capabilities.

Also remove the "deployed" tracking from the JaccEarDUP as that is not
per-depoyment. The undeploy method is no-op anyway but we should leave
the call in place in case we want to add further clean up.

bstansberry reviewed (approved):
@darranl LGTM. If you can put a bit more detail in the WFLY description that would be good.
